src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+                                                                      const ChunkID chunk_id,
+                                                                      const std::string data_type) const {
+  // determine the set of rows that should be included in the scan output,
+  // i.e. find the row indexes that match the filter condition.
+
+  // accumulate the indexes of the rows that should be included in the output in include_rows_ptr.
+  // The container is initialized later on when we perform the actual filtering.
+  // This declaration is needed so that we can pass the object outside of the
+  // scope of the lambda below.
+  std::shared_ptr<std::vector<ChunkOffset>> include_rows_ptr;
+
+  // get segment that we want to filter on and cast it to the right type. Then,
+  // perform a scan on the segment based on the segment type (value/dict/reference).
+  auto segment_ptr = chunk_ptr->get_segment(_column_id);
+  resolve_data_type(data_type, [&](auto type) {
+    using Type = typename decltype(type)::type;
+    // case 1: segment is value segment
+    const auto typed_value_segment_ptr = std::dynamic_pointer_cast<ValueSegment<Type>>(segment_ptr);
+    if (typed_value_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(typed_value_segment_ptr);
+      return;
+    }
+    // case 2: segment is dictionary segment
+    const auto typed_dict_segment_ptr = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment_ptr);
+    if (typed_dict_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(typed_dict_segment_ptr);
+      return;
+    }
+    // case 3: segment is reference segment
+    const auto ref_segment_ptr = std::dynamic_pointer_cast<ReferenceSegment>(segment_ptr);
+    if (ref_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(ref_segment_ptr);
+      return;
+    }
+
+    // we do not support any segment types beyond value, dict, and reference segments
+    throw std::runtime_error("unrecognized segment class at chunk id " + std::to_string(chunk_id) + " and column id " +
+                             std::to_string(_column_id));
+  });
+
+  include_rows_ptr->shrink_to_fit();
+  return include_rows_ptr;
+}
+
+const std::shared_ptr<Chunk> TableScan::subset_chunk(
+    const std::shared_ptr<const Table> table_ptr, const std::shared_ptr<const Chunk> chunk_ptr, const ChunkID chunk_id,
+    const std::shared_ptr<std::vector<ChunkOffset>>& include_rows_ptr) const {
+  // create a chunk that consists of reference segments that point
+  // only to the values that we want to include in the output table.
+  // The values that we want to include are given in include_rows_ptr.
+
+  // accumulate the new reference segments in this chunk.
+  auto out_chunk_ptr = std::make_shared<Chunk>();
+
+  // we convert each segment to a reference segment independently
+  auto n_segments = chunk_ptr->column_count();
+  for (auto col_id = ColumnID{0}; col_id < n_segments; ++col_id) {
```

Comment:
You are building the `pos_list` inside this loop, however I do not think that it will ever be different for the columns. Hence, you could create the `pos_list` only once and add it to a `vector<ReferenceSegment>`

Reply:
Important note!
This could bring great performance increases.